4.0 Limitations of security (#637)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/ql/administration/security/security-and-labels.asciidoc

=== How does role-based access control impact label visibility

> ==== Traversing the graph with multi-labeled nodes

Reviewer: I don't really like that you have 2 headlines after each other here. With the `5.5.5.2. Security and labels`  headline, that means 3 headlines in a row. I think there are too many levels of sections even without this, plus that it looks a bit like there is missing text between the headlines. 

Perhaps you could merge "Security and labels" and "How does role-based access control impact label visibility" into one single headline and then change the other headlines of the section to have 3 instead of 4 ===?

The general influence of access control privileges on graph traversal is described in detail in <<administration-security-subgraph, Graph and sub-graph access control>>.
The following section will only focus on nodes because of their ability to have multiple labels. Relationships can only ever have one type
and thus they do not exhibit the behavior this section aims to clarify.

> While this document will not mention relationships further, the general function of the traverse privilege also applies to them.

Reviewer 2: ```suggestion
While this section will not mention relationships further, the general function of the traverse privilege also applies to them.
```

For any node that is traversable, due to `GRANT TRAVERSE` or `GRANT MATCH`,
the user can get information about the labels attached to the node by calling the built-in `labels()` function.
In the case of nodes with multiple labels, this can seemingly result in labels being returned to which the user
wasn't directly granted access to.

To give an illustrative example, imagine a graph with three nodes: one labeled `:A`, one labeled `:B` and one with `:A :B`.

> We also have user with the role "custom" as defined by:

Reviewer: Maybe the role name "custom" should be backticked instead of in quotes?

>     GRANT TRAVERSE ON GRAPH * NODES A TO custom;

Reviewer 2: Code block

If that user were to execute

>     MATCH (n:A) RETURN n, labels(n);

Reviewer 2: Code block

they would be returned two nodes: the node that was labeled with `:A` and the node with labels `:A :B`.

In contrast, executing

>     MATCH (n:B) RETURN n, labels(n);

Reviewer 2: Code block

> will return only the one node that has both labels: `:A :B`. Even though `:B` was not whitelisted for traversal, there is one

Reviewer 2: See other comment about black/whitelisting
```suggestion
will return only the one node that has both labels: `:A :B`. Even though `:B` was not allowlisted for traversal, there is one
```
Alternatively, "..was not allowed access for traversal..."

> will return only the one node that has both labels: `:A :B`. Even though `:B` was not whitelisted for traversal, there is one

Reviewer 3: ```suggestion
will return only the one node that has both labels: `:A :B`. Even though `:B` was not allowed access for traversal, there is one
```

> node with that label accessible in the data because of the whitelisted label `:A` that is attached to the same node.

Reviewer 2: As above
```suggestion
node with that label accessible in the data because of the allowlisted label `:A` that is attached to the same node.
```

// Get A,B but was only granted A

> If a user is denied traverse on a label they will never get results from any node that has this label (also)

Reviewer: I think the `(also)` here doesn't provide any value, it just make it harder to read.

attached to it. Thus, the label name will never show up for them. For our example this can be done by executing:

>     DENY TRAVERSE ON GRAPH * NODES B TO custom;

Reviewer 2: Code block

The query

>     MATCH (n:A) RETURN n, labels(n);

Reviewer 2: Code block

will now return the node only labeled with `:A`, while the query

>     MATCH (n:B) RETURN n, labels(n);

Reviewer 2: Code block

will now return no nodes.

==== The db.labels() procedure

> In contrast to the normal graph traversal described in the previous section the built-in `db.labels()` procedure

Reviewer 2: ```suggestion
In contrast to the normal graph traversal described in the previous section, the built-in `db.labels()` procedure
```

is not processing the data graph itself but the security rules defined on the system graph.
That means:

* if a label is explicitly whitelisted (granted), it will be returned by this procedure.

> * if a label is blacklisted (denied) or isn't explicitly whitelisted it will not be returned by this procedure.

Reviewer 2: As above
```suggestion
* if a label is denied or isn't explicitly allowed it will not be returned by this procedure.
```

To reuse the example of the previous section: imagine a graph with three nodes: one labeled `:A`, one labeled `:B` and one with `:A :B`.

> We also have user with the role "custom" as defined by:

Reviewer: Maybe the role name "custom" should be backticked instead of in quotes?

>     GRANT TRAVERSE ON GRAPH * NODES A TO custom;

Reviewer 2: Code block

> This means that only label `:A` is explicitly whitelisted.

Reviewer 2: ```suggestion
This means that only label `:A` is explicitly allowlisted.
```

Thus, executing

>     CALL db.labels();

Reviewer 2: Code block

wasn't directly granted access to.

To give an illustrative example, imagine a graph with three nodes: one labeled `:A`, one labeled `:B` and one with `:A :B`.

> We also have user with a role `custom` as defined by:

Reviewer: ```suggestion
We also have a user with a role `custom` as defined by:
```

* if a label is denied or isn't explicitly allowed it will not be returned by this procedure.

To reuse the example of the previous section: imagine a graph with three nodes: one labeled `:A`, one labeled `:B` and one with `:A :B`.

> We also have user with a role `custom` as defined by:

Reviewer: ```suggestion
We also have a user with a role `custom` as defined by:
```
